### PR TITLE
[rpc/middleware][consul] plumb intercept off, add server level happy test

### PR DESCRIPTION
**Pull request by FFMMM**
### overview

This patch does 2 main things:

* plumbs down a config `GetNetRPCInterceptorFunc` option (not exposed thru to the agent)
* uses that func to add server level happy path testing AND prove that we can turn the `net/rpc` interceptor off if we ever need to

Additionally, it adds two small changes to the `rpc/middleware` package:

* exports the recorder func
* make `interceptor_test.go` run in parallel if needed
 

### reviewer notes

* best viewed by commits

### CODE REVIEW

File: agent/consul/server_test.go
Comment (FFMMM) [inline agent/consul/server_test.go:1235]: * see `interceptors_test.gp` for "data" level and unit level testing
* see `metrics_test.go` for e2e/ bdd testing

all in main already

File: agent/consul/server_test.go
Comment (mkeeler) [inline agent/consul/server_test.go:1166]: Is there a data race here? I assume the test agent behind the scenes will be performing RPCs and swapping in an RPC recorder seems like it could cause the test to panic in the worst case.

File: agent/consul/config.go
Comment (mkeeler) [inline agent/consul/config.go:124]: I don't think I have an opinion either way but did you consider putting this in the Deps instead of the Config?

*Most* of whats in Config are simple values and not types/funcs (one exception being the `NotifyListen` func above here).

Then again, I don't know how applicable it is with what else is in Deps. What do you think?

File: agent/consul/server.go
Comment (dhiaayachi) [inline agent/consul/server.go:380]: Maybe inject the recorder to the server by adding it as a param to NewServer. This way we can avoid the race that @mkeeler is referring to.

File: agent/consul/server.go
Comment (dhiaayachi) [inline agent/consul/server.go:395]: What are the implication of creating a new rpc server in here and then creating a new one later, do the NewServer() call use any system resources (ex socket)?
I think another way in here is to make `GetNetRPCInterceptorFunc` return a nil (instead of being nil) in the case we don't want to intercept the calls and make sure that the nil interceptor is a no op.

File: agent/consul/server_test.go
Comment (dhiaayachi) [inline agent/consul/server_test.go:1147]: maybe just fail the test directly in here, we don't expect any recording to happen in here right?

**Review by dhiaayachi — CHANGES_REQUESTED** [review]
Thank you for working on this @FFMMM. The changes in here make the tests for rpc metrics definitely easier to write and reason about.

I had a few comments around the recorder creation in `NewServer` and the rpc server creation code.

File: agent/consul/config.go
Comment (FFMMM) [reply agent/consul/config.go:124]: That's a fair point. Chatting with Dhia more, I want to move this to `Deps` and I think it should work better ✍🏼

File: agent/consul/server_test.go
Comment (FFMMM) [reply agent/consul/server_test.go:1147]: that's a good point -- I recall that folks were not inclined to do this when I was adding interceptor;

i'm slightly biased to pretend that this function does some recording and then to look at the `map` to see that the right values are in there.

File: agent/consul/server_test.go
Comment (FFMMM) [reply agent/consul/server_test.go:1166]: ‼️ that is quite possible; I'm moving the `recorder` to be a part of deps too so this shouldn't be a problem anymore.

**Comment by FFMMM** [comment]
@dhiaayachi, I addressed the changes requested. It seems that `verify-ci` is currently stuck so I am going to dry to dismiss the review to see if it kicks the tires on the workflow.

**Comment by FFMMM** [comment]
or maybe it's the outage 🙈  https://status.circleci.com/

**Review by dhiaayachi — APPROVED** [review]
LGTM with a small nit!

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/627963.

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/628205.

**Comment by hc-github-team-consul-core** [comment]
:cherries::x: Cherry pick of commit 5245251bbf73b20943eec4d4f8dd5038916cd3f8 onto `release/1.12.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/628205)

**Pull request merged**
